analyzer/src/types/ctx.rs

```diff
@@ -0,0 +1,80 @@
+use crate::relations::{Relations, SourceObjectId, Symbol};
+use crate::types::hir::TypeId;
+use crate::types::{BOOL, FLOAT, INT, NOTHING, STRING};
+use std::collections::HashMap;
+
+/// Holds the symbol to type mapping.
+///
+/// The actual type definition is in the [`crate::types::Typing`] struct.
+pub struct TypeContext {
+    names: HashMap<String, TypeId>,
+    locals: HashMap<SourceObjectId, Vec<TypeId>>,
+}
+
+impl TypeContext {
+    pub(crate) fn new() -> Self {
+        Self {
+            names: HashMap::new(),
+            locals: HashMap::new(),
+        }
+    }
+
+    pub(crate) fn lang() -> Self {
```

Comment:
this also feels wrong as the context is meant to mutate, i think it'd be better to fill the lang types on a new instance instead of defining a `lang` constructor, or refactor this with `with_lang_prefilled` or something else

Reply:
It is a default constructor, that explicitly states that it initially contains the definition for primitive types. Creating a really empty `TypeContext` is something that should be disallowed, since most of the consumers expect the primitives types to always be present. See the constants that always points to the same index.
This context will be mutated, but already inserted types will never be removed. Merging this constructor with the `fill_lang` method would be the best move in my opinion.

Reply:
the definition of `TypeContext#lang` in this way lets think that the context is bound to the lang types, the primitives. If i mutate it (i.e push a new type into it), it can feels like i'm creating a new lang type, which isn't, i agree with the fact that we cannot create a type context without the lang types, thus indeed hiding the `fill_lang` inside `new` or creating a `with_lang_prefilled` can be a good idea